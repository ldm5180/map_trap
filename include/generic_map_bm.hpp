#ifndef GENERIC_MAP_BM_HPP
#define GENERIC_MAP_BM_HPP

#include "add_to_map.hpp"

#include <boost/hana.hpp>
#include <cstdlib>

static constexpr auto has_reserve =
    boost::hana::is_valid([](auto &&x) -> decltype((void)x.reserve(0)) {});

template <class map_t> static void BM_map_insert(benchmark::State &state) {
  std::srand(42);
  map_t map{};
  if
    constexpr(has_reserve(map)) { map.reserve(state.range(0)); }

  for (auto _ : state) {
    add_to_map(state.range(0), map);
  }
}

template <class map_t> static void BM_map_iterate(benchmark::State &state) {
  std::srand(42);
  map_t map{};
  if
    constexpr(has_reserve(map)) { map.reserve(state.range(0)); }

  for (auto i = 0; i < state.range(0); ++i) {
    add_to_map(state.range(0), map);
  }

  for (auto _ : state) {
    // Nothing to do to sort them. Can simply iterate.
    typename map_t::mapped_type v{};
    for (const auto &kv : map) {
      benchmark::DoNotOptimize(v = kv.second);
    }
  }
}

template <class map_t> static void BM_map_find(benchmark::State &state) {
  std::srand(42);
  map_t map{};
  if
    constexpr(has_reserve(map)) { map.reserve(state.range(0)); }
  std::vector<typename map_t::key_type> keys{};
  keys.reserve(state.range(0));

  for (auto i = 0; i < state.range(0); ++i) {
    add_to_map_and_key_list(state.range(0), map, keys);
  }

  for (auto _ : state) {
    typename map_t::mapped_type v{};
    const auto key = keys[std::rand() % keys.size()];
    benchmark::DoNotOptimize(v = map[key]);
  }
}

#endif // GENERIC_MAP_BM_HPP
