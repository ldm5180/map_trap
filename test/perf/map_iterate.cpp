#include "generic_map_bm.hpp"

#include <benchmark/benchmark.h>
#include <map>

BENCHMARK_TEMPLATE(BM_map_iterate, std::map<int64_t, int64_t>)
    ->RangeMultiplier(2)
    ->Range(8, 8 << 4);
BENCHMARK_MAIN();
