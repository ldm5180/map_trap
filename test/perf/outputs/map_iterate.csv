Run on (4 X 2594.09 MHz CPU s)
2017-10-28 17:23:59
name,iterations,real_time,cpu_time,time_unit,bytes_per_second,items_per_second,label,error_occurred,error_message
"BM_map_iterate<std::map<int64_t, int64_t>>/8",2387241,279.016,279.001,ns,,,,,
"BM_map_iterate<std::map<int64_t, int64_t>>/16",613415,1153.13,1153.11,ns,,,,,
"BM_map_iterate<std::map<int64_t, int64_t>>/32",103588,6476.31,6475.21,ns,,,,,
"BM_map_iterate<std::map<int64_t, int64_t>>/64",16843,40074.4,40071.5,ns,,,,,
"BM_map_iterate<std::map<int64_t, int64_t>>/128",2390,259004,259001,ns,,,,,
