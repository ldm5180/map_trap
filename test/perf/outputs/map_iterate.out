Run on (4 X 2594.09 MHz CPU s)
2017-10-28 16:30:02
name,iterations,real_time,cpu_time,time_unit,bytes_per_second,items_per_second,label,error_occurred,error_message
"BM_map_iterate<std::map<int, std::string>>/8",713520,902.463,902.626,ns,,,,,
"BM_map_iterate<std::map<int, std::string>>/16",168997,4100.97,4102.99,ns,,,,,
"BM_map_iterate<std::map<int, std::string>>/32",39032,18274.7,18283.7,ns,,,,,
"BM_map_iterate<std::map<int, std::string>>/64",7211,91646.6,91692.5,ns,,,,,
"BM_map_iterate<std::map<int, std::string>>/128",1667,394343,394541,ns,,,,,
